### DISCUSSION TITLE
Svelte `bind` equivalent

### DISCUSSION THREAD

**Discussion by Lebenoa** [leptos-rs/leptos#337] [Q&A] [answered]
Hello! Sorry for this dumb question. I've never use `React` but `Svelte`. And I wonder how to implement something like this in `leptos`  

I tried reading `React` docs but that did not help me either.

## Say for example, in Svelte
```svelte
<label class="form-control">
    <input
        type="checkbox"
        bind:checked={myBoolean}
     />
    Check
</label>
```

## For value
```svelte
<input
    type="number"
    use:validator={myNumber}
    bind:value={myNumber}
    min="0"
/>
```

The `use` directive is fine. I figured out how but `bind`

### P.S. Sorry for my bad English. I'm not a native speaker of English.

**Comment by gbj** [comment, marked as answer]
No worries. This framework is fairly different from React but quite similar to SolidJS, so you may have more luck there. In either case, you want to look up React or Solid "controlled components" and "uncontrolled components" for patterns.

```rust
let (name, set_name) = create_signal(cx, "gbj".to_string());
view! { cx,
  <input type="text"
    // updates the `value` property in the DOM reactively to match `name`
    prop:value=name
    // event_target_value is the equiv. of JS event.target.value
    // just a helper to smooth out some Rust/JS type awkwardness
    on:input=move |ev| set_name(event_target_value(&name))
  />
}
```

The equivalent goes for `prop:checked` and `event_target_checked` on a checkbox.

Basically:
1) declare a signal
2) on the relevant DOM, update the signal
3) feed the signal into the value property in the DOM

**Reply by Lebenoa** [reply]
Hello! Thanks for the reply! I can get `checkbox` to work with this
```rust
on:change=move |ev| setter(event_target_checked(&ev)) // `setter` is `WriteSignal<bool>` and parse the `Event` instead of `ReadSignal`
```

but I can't get it working with `input` neither **text** or **number**  
Maybe I just need reactive state like this in Svelte

```svelte
<script>
	let name = 'world';
	let someNumber = 0;
	
	$: if (name == 'world') {
		name = 'earth'
	}
	
	$: if (someNumber > 1500) {
		someNumber = 1500
	} else if (someNumber < 0) {
		someNumber = 0
	} 
</script>

<input type="text" bind:value={name} />
{name}

<br />

<input type="number" bind:value={someNumber} />
{someNumber}
```
Where you type `world` in textbox, it'll replace `world` with `earth`  
and when you put `someNumber` above **1500** (e.g. 1600) it'll replace that with **1500**

**Demo**: https://svelte.dev/repl/5da744d46e0142d6aadb276f024b0d21?version=3.55.1

How can I do something like this? (Sorry if this is such a common question. I'm still not fully wake up yet)

**Reply by luckynumberke7in** [reply]
Hi Meonako, 

So a signal is a reactive bit of state based on solid-js. You would view the state using 'name' and set a new value for 'name' with 'set_name'. If you want to make a check and mutate the data based on specific criteria, you can use 'create_effect'. The 'create_effect' function will re-run whenever one or more of the signals values you put inside of it changes.

Something like the 3rd example in this doc: [Leptos reactivity](https://github.com/leptos-rs/leptos/blob/c8f0988e53d2521abe4ec424502c5a7600495f81/docs/book/src/04_reactivity.md)

Leptos is still quite new and working on a website / getting docs sorted out, but for a curated example of the core concepts of Leptos & reactivity you can see [Solid JS website](https://www.solidjs.com/tutorial/introduction_basics) for now!

**Reply by Lebenoa** [reply]
Thank you for your help! but I'm still not quite there yet...

https://user-images.githubusercontent.com/76484203/213360930-3a66f0b5-41e0-47c5-9c0b-059ca3c22d04.mp4

When hitting `11111` for the first time, it replace that with `1500` but after that it just ignored.  

Here's my component code
```rust
#[component]
pub fn Input(
    cx: Scope,

    /// Read Signal. (Current Value)
    reader: ReadSignal<i32>,
    
    /// Write Signal.
    setter: WriteSignal<i32>,
    
    /// Max Value.
    max_value: i32,

) -> impl IntoView {
    let output = move || reader();
    create_effect(cx, move |_| {
        if output() > max_value {
            setter(max_value)
        } else if output() < 0 {
            setter(0)
        }
        log::info!("effect triggered: {}", output());
    });

    let input_listener = move |ev| {
        let target_value = event_target_value(&ev);

            if target_value == "" {
                setter(0);
            } else if target_value.parse::<i32>().unwrap() > max_value {
                setter(max_value)
            } else { 
                setter(target_value.parse().unwrap());
            };
        // setter(value)
    };

    view! {cx,
        <input type="number"
            prop:value=output
            on:input=input_listener
            on:change=input_listener
        />
        "/ " {max_value}
    }
}
```

**Reply by Lebenoa** [reply]
Never mind! I've decided to write in JS instead. Thanks for your help!

**Reply by gbj** [reply]
@Meonako Just for the sake of clarity, your example doesn't show it being ignored -- it shows 1500 being logged with each additional 1. It's just that your console is set to collapse duplicate logs, so there's a little 1, 2, 3 next to the log each time you're pressing a key. Unless there's something I'm not understanding about the example it's working exactly as you intend it.

**Reply by Lebenoa** [reply]
Ah I mean I can let this Svelte example explain this for me ig

https://user-images.githubusercontent.com/76484203/213458217-842dcc77-dd87-409c-bf4f-5777cfc7655f.mp4

If I keep pressing 1 in this Svelte example, **THE INPUT VALUE** stay at **1500** while leptos goes `1500111...`

Thanks again for the reply :)

**Reply by gbj** [reply]
Ahhhh okay I see. This is pretty interesting. Here's a version that works as the Svelte one does. (I edited the component slightly so I could play around with it, but I'm sure you can see the difference)

This is basically using an "uncontrolled" component with a `NodeRef` rather than a "controlled" component with `prop:value`, but using an effect to call `HtmlInputElement::set_value` directly. I'm not sure why calling `set_value` here has a different effect from setting the `value` prop, but... there you are. 

```rust
#[component]
pub fn Input(
    cx: Scope,

    /// Max Value.
    max_value: i32,
) -> impl IntoView {
    let (reader, setter) = create_signal(cx, 0);
    let input_node = NodeRef::<HtmlElement<Input>>::new(cx);

    create_effect(cx, move |_| {
        leptos::log!("value is {}", reader.get());
        if let Some(node) = input_node.get() {
            node.set_value(&reader.get().to_string());
        }
    });

    let input_listener = move |ev| {
        let target_value = event_target_value(&ev);

        if target_value == "" {
            setter(0);
        } else if target_value.parse::<i32>().unwrap() > max_value {
            setter(max_value)
        } else {
            setter(target_value.parse().unwrap());
        };

    };

    view! {cx,
        <input type="number"
            node_ref=input_node
            on:input=input_listener
        />
        "/ " {max_value}
    }
}
```

**Reply by Lebenoa** [reply]
Hey! Thank you so much! I am extremely grateful for your help. It's working now!
